fal/models/team.py

```diff
@@ -62,12 +89,66 @@ def bench_swap(self, active_anime: Anime, bench_anime: Anime, week: int) -> None
             .all()
         )
 
-        assert len(this_week_anime) == 2
+        assert len(this_week_anime_involved) == 2
 
-        for anime in this_week_anime:
+        for anime in this_week_anime_involved:
             if anime.anime_id == active_anime._entity.id:
                 assert not anime.bench
                 anime.bench = 1
             elif anime.anime_id == bench_anime._entity.id:
                 assert anime.bench
                 anime.bench = 0
+
+        successful_swap = orm.BenchSwap(
+            team_id=self._entity.id,
+            week=week,
+            to_bench=active_anime._entity.id,
+            from_bench=bench_anime._entity.id,
+        )
+        self._session.add(successful_swap)
+        self.commit()
+
+    def add_anime_to_team(self, anime: Anime, bench: bool = False) -> None:
+        """
+        Should only be called in week 0, adds anime to team's active or bench.
+
+        Raises sqlalchemy.exc.IntegrityError if anime already exists on team.
+        """
+        first_week_anime = orm.TeamWeeklyAnime(
+            team_id=self._entity.id,
+            anime_id=anime._entity.id,
+            week=0,
+            bench=int(bench),
+        )
+        self._session.add(first_week_anime)
+        self.commit()
+
+    def get_anime(self, week: int) -> WeekSnapshotOfTeamAnime:
+        """
+        Returns all anime on team during week specified
+        """
+        this_week_anime = (
+            self._session.query(orm.TeamWeeklyAnime)
+            .filter(
+                orm.TeamWeeklyAnime.team_id == self._entity.id,
+                orm.TeamWeeklyAnime.week == week,
+            )
+            .all()
+        )
+
+        snapshot = WeekSnapshotOfTeamAnime(week=week)
+        for team_weekly_anime in this_week_anime:
+            anime = Anime.from_orm_anime(team_weekly_anime.anime, self._session)
+            if team_weekly_anime.bench:
+                snapshot.bench.append(anime)
+            else:
+                snapshot.active.append(anime)
+
+        return snapshot
+
+
+@attr.s(auto_attribs=True)
+class WeekSnapshotOfTeamAnime(object):
```

Comment:
Maybe this isn't true anymore because of attrs, but I thought classes in Python need to be defined above where they're used in the code.

Reply:
they do if they're used. because Python is a dynamic language though, if the usage is inside a function, then the class doesn't need to be defined until the function is called, because that would be the first time the function is used.

this is in contrast to a statically compiled language like c++ where during compiling of the code, the function would link the class symbol to a forward declaration

Reply:
That makes sense, thanks for explaining.